base_layer/core/src/base_node/state_machine_service/handle.rs

```diff
@@ -0,0 +1,90 @@
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::base_node::state_machine_service::states::{StateEvent, StatusInfo};
+use core::sync::atomic::{AtomicBool, Ordering};
+use futures::{stream::Fuse, StreamExt};
+use std::sync::Arc;
+use tari_broadcast_channel::Subscriber;
+use tari_shutdown::ShutdownSignal;
+use tokio::sync::watch;
+
+#[derive(Clone)]
+pub struct StateMachineHandle {
+    state_change_event_subscriber: Subscriber<StateEvent>,
+    status_event_receiver: watch::Receiver<StatusInfo>,
+    shutdown_signal: ShutdownSignal,
+    initial_blockchain_sync_success: Arc<AtomicBool>,
+}
+
+impl StateMachineHandle {
+    pub fn new(
+        state_change_event_subscriber: Subscriber<StateEvent>,
+        status_event_receiver: watch::Receiver<StatusInfo>,
+        initial_blockchain_sync_success: Arc<AtomicBool>,
+        shutdown_signal: ShutdownSignal,
+    ) -> Self
+    {
+        Self {
+            state_change_event_subscriber,
+            status_event_receiver,
+            initial_blockchain_sync_success,
+            shutdown_signal,
+        }
+    }
+
+    /// This clones the receiver end of the channel and gives out a copy to the caller
+    /// This allows multiple subscribers to this channel by only keeping one channel and cloning the receiver for every
+    /// caller.
+    pub fn get_state_change_event_stream(&self) -> Subscriber<StateEvent> {
+        self.state_change_event_subscriber.clone()
+    }
+
+    pub fn get_state_change_event_stream_fused(&self) -> Fuse<Subscriber<StateEvent>> {
+        self.get_state_change_event_stream().fuse()
+    }
+
+    /// This clones the receiver end of the channel and gives out a copy to the caller
+    /// This allows multiple subscribers to this channel by only keeping one channel and cloning the receiver for every
+    /// caller.
+    pub fn get_status_event_stream(&self) -> watch::Receiver<StatusInfo> {
+        self.status_event_receiver.clone()
+    }
+
+    pub fn get_status_event_stream_fused(&self) -> Fuse<watch::Receiver<StatusInfo>> {
+        self.get_status_event_stream().fuse()
+    }
+
+    pub fn shutdown_signal(&self) -> ShutdownSignal {
+        self.shutdown_signal.clone()
+    }
+
+    /// This lets any interested party update the the initial blockchain sync status
+    pub fn set_initial_blockchain_sync_success_status(&self, status: bool) {
```

Comment:
I'm not sure about this, but I'll have to see where it's used. Seems like this field should only ever be readonly and determined by the base node

Reply:
I agree with this, was intended to be only read-only in the handle. Will remove.